File: backend/application-service/src/main/java/com/gl/application_service/service/ApplicationService.java
package com.gl.application_service.service;

import com.gl.application_service.dto.ApplicationDTO;
import com.gl.application_service.enums.Status;
import com.gl.application_service.exception.ApplicationNotFoundException;

import java.util.List;

public interface ApplicationService {
    void createApplication(ApplicationDTO applicationDTO) throws ApplicationNotFoundException;
    List<ApplicationDTO> getAllApplicationsOfProject(Long projectId) throws ApplicationNotFoundException;
    List<ApplicationDTO> getAllApplicationsOfUser(Long userId) throws ApplicationNotFoundException;
    void updateStatus(Long applicationId, Status status) throws ApplicationNotFoundException;
    void deleteApplication(Long applicationId) throws ApplicationNotFoundException;
    void deleteAllApplicationsForProject(Long projectId);

}

File: frontend/src/api/applicationService.ts
import axiosClient from "./axiosClient";

export enum ApplicationStatus {
  PENDING = "PENDING",
  ACCEPTED = "ACCEPTED",
  REJECTED = "REJECTED",
}

export interface ApplicationDTO {
  applicationId?: number;
  projectId: number;
  userId: number;
  status: ApplicationStatus;
  appliedAt: string;
  updatedAt: string;
}

export const applicationService = {
  // @PostMapping
  createApplication: async (data: Partial<ApplicationDTO>): Promise<string> => {
    const response = await axiosClient.post("/api/applications", data);
    return response.data;
  },

  // @GetMapping("/project/{projectId}")
  getApplicationsByProject: async (
    projectId: number,
  ): Promise<ApplicationDTO[]> => {
    const response = await axiosClient.get(
      `/api/applications/project/${projectId}`,
    );
    return response.data;
  },

  // @GetMapping("/user/{userId}")
  getApplicationsByUser: async (userId: number): Promise<ApplicationDTO[]> => {
    const response = await axiosClient.get(`/api/applications/user/${userId}`);
    return response.data;
  },

  // @PutMapping("/{applicationId}/status")
  // Note: status is passed as a @RequestParam in your backend
  updateStatus: async (
    applicationId: number,
    status: ApplicationStatus,
  ): Promise<string> => {
    const response = await axiosClient.put(
      `/api/applications/${applicationId}/status`,
      null,
      {
        params: { status },
      },
    );
    return response.data;
  },
};
